Introduce SecurityContext

fickludd opened the pull request:
The `SecurityContext` separates `AuthSubject` and `AccessMode` from eachother, by using composition rather than inheritance. One benefit of this refactoring is the removal of numerous type checks and casts that we do to cast `AccessMode`s to `AuthSubject`s.

fickludd committed:
Introduced SecurityContext that replaces AccessMode

fickludd committed:
Moved securityContext enums until new AnonymousContext class

fickludd committed:
Contain auth subject in security context

- Auth subject does no longer inherit security context
- CouldBeAdmin interface to check if admin in enterprise

fickludd committed:
Rename `Allowance` back to `AccessMode`

fickludd committed:
Refactor the way we freeze SecurityContext

This enables the enterprise classes to get the correct types.
Also fixes a test that did not have the correct assertions.

henriknyman commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -229,7 +229,7 @@ public KernelTransactionImplementation initialize(
         this.transactionEvent = tracer.beginTransaction();
         assert transactionEvent != null : "transactionEvent was null!";
-        this.accessMode = accessMode.getSnapshot();
+        this.securityContext = securityContext.freeze();
```
It would be nicer if the transaction could keep the AccessMode member as it was, and do like this.accessMode = securityContext.getAccessModeSnapshot(). That would also minimize the necessary code changes necessary. If we need to get SecurityContext from transaction it can just be an additional member.

henriknyman commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -796,11 +796,11 @@ public long getCommitTime()
 
     @Override
-    public Revertable overrideWith( AccessMode mode )
+    public Revertable overrideWith( SecurityContext context )
```
...then this could be unchanged (overrideWith( AccessMode mode ))

henriknyman commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java:
```diff
@@ -1494,59 +1495,67 @@ public void unregisterExecutingQuery( ExecutingQuery executingQuery )
     public RawIterator<Object[], ProcedureException> procedureCallSchemaOverride( QualifiedName name, Object[] input ) throws ProcedureException
     {
-        return callProcedure( name, input, new OverriddenAccessMode( tx.mode(), AccessMode.Static.FULL ) );
+        return callProcedure( name, input, new OverriddenAccessMode( tx.securityContext().mode(), AccessMode.Static.FULL ) );
```
...all these could remain unchanged tx.mode()

henriknyman commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java:
```diff
@@ -1597,7 +1609,7 @@ private Object callFunction(
         statement.assertOpen();
 
-        try ( KernelTransaction.Revertable ignore = tx.overrideWith( mode ) )
+        try ( KernelTransaction.Revertable ignore = tx.overrideWith( tx.securityContext().freeze( mode ) ) )
```
What is freeze() doing here?

fickludd replied:
It's just a way to associate the same subject with a new mode (within a SecurityContext). We'll change this to `withMode` to make it more clear.

henriknyman commented on community/server/src/main/java/org/neo4j/server/rest/dbms/UserService.java:
```diff
@@ -146,15 +146,11 @@ public Response setPassword( @PathParam("username") String username, @Context Ht
     }
 
-    private AuthSubject getSubjectFromPrincipal( Principal principal )
+    private SecurityContext getSecurityContextFromPrincipal( Principal principal )
```
This looks like a duplicate of the static method getSecurityContextFromUserPrincipal() above

fickludd replied:
Removed this

henriknyman commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/PersonalUserManager.java:
```diff
@@ -34,14 +35,16 @@
+    private EnterpriseUserManager userManager;
+    private SecurityContext securityContext;
+    private AuthSubject authSubject;
+    private SecurityLog securityLog;
```
Why was final removed?

fickludd replied:
Merge mistake. Fixed.

henriknyman commented on enterprise/security/src/main/resources/simplelogger.properties:
```diff
@@ -0,0 +1,34 @@
+
+# Set to true if you want the last component of the name to be included in output messages.
+# Defaults to false.
+#org.slf4j.simpleLogger.showShortLogName=false
```
Was this added by mistake? We are not using ehcache..

fickludd replied:
Oops, removed.

henriknyman commented on community/kernel/src/main/java/org/neo4j/kernel/api/security/SecurityContext.java:
```diff
@@ -0,0 +1,105 @@
+
+    default String defaultString( String name )
+    {
+        return String.format( "%s{ securityContext=%s, allowance=%s }", name, subject().username(), mode() );
```
allowance -> accessMode ?

henriknyman replied:
securityContext -> username ?

fickludd replied:
yes

henriknyman reviewed (changes requested):
Looks good in general

henriknyman commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/ExecutionEngine.scala:
```diff
@@ -150,8 +150,8 @@ class ExecutionEngine(val queryService: GraphDatabaseQueryService, logProvider:
         // Temporarily change access mode during query planning
-        // NOTE: This will force read access even if the current transaction did not have it
-        val revertable = tc.restrictCurrentTransaction(AccessMode.Static.READ)
+        // NOTE: This will force read allowance if the current transaction did not have it
```
allowance -> access

henriknyman commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_1/TransactionBoundQueryContext.scala:
```diff
@@ -590,8 +590,8 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
-    val mode = transactionalContext.accessMode
-    allowed.nonEmpty && !mode.isOverridden && mode.getOriginalAccessMode.allowsProcedureWith(allowed)
+    val allowance = transactionalContext.securityContext.mode()
+    allowed.nonEmpty && !allowance.isOverridden && allowance.allowsProcedureWith(allowed)
```
allowance -> mode

fickludd committed:
Minor PR feedback fixes.

henriknyman reviewed (commented):
Minor traces of allowance left.. :)

henriknyman commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/StandardEnterpriseSecurityContext.java:
```diff
@@ -0,0 +1,228 @@
+                shiroSubject.isAuthenticated() && shiroSubject.isPermitted( SCHEMA_READ_WRITE ),
+                shiroSubject.getAuthenticationResult() == AuthenticationResult.PASSWORD_CHANGE_REQUIRED,
+                authManager.getAuthorizationInfo( shiroSubject.getPrincipals() )
+            );
```
Will this ever be called outside except from freeze()  ? Maybe we could use a different name for the implementation method to make the code more clear.

henriknyman replied:
I guess the place where we have the opportunity to use a flyweight has now moved to here, but now authorization info is also tied to it. I wonder if that should be generalized as a set of roles to make AccessMode more agnostic to Apache Shiro.

Mats-SX replied:
I don't know how exactly to act on these comments. Is it OK if we perhaps address them in an upcoming commit?

fickludd committed:
Change error message for PERMISSION_DENIED

We augment the error message on permission denied to detail both the
AuthSubject (when available) and the full AccessMode, including roles in enterprise.

henriknyman commented on community/kernel/src/main/java/org/neo4j/kernel/api/security/AnonymousContext.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Override
+    public SecurityContext freeze( AccessMode mode )
```
I guess you meant that this will be renamed to `withMode` ? That would make it much more clear.

henriknyman reviewed (commented):
After a second look I think my initial concern was more about the immutable design of security context snapshotting and if that would lead to more object creation than if access mode overriding would just mutate the access mode within the security context. I was imagining that the refactoring would open up the possibility that combined with a flyweight pattern for access modes, it would make it possible to have tight loops of procedure calls without generating much garbage. Maybe my concern is exaggerated, and there are other pros of an immutable design. It would be interesting to do some micro benchmarks on procedure calls though.

fickludd committed:
Minor cleaup

craigtaverner commented on community/bolt/src/test/java/org/neo4j/bolt/security/auth/BasicAuthenticationTest.java:
```diff
@@ -191,42 +208,19 @@ public void shouldThrowWithNoScheme() throws Exception
-    }
-
-    @Test
-    public void shouldFailOnMalformedToken() throws Exception
```
Why was this test removed?

Mats-SX replied:
I don't know. I'm reinstating it.

boggle commented:
This looks good overall. Thanks for picking this up! 

I'd like to suggest to add more extensive docs to `SecurityContext` (class level and methods), possibly to `AccessMode` as well (didn't see that one pass by). Naming wise, isn't `AccessMode` more like `Permissions`? Also perhaps rename `Frozen` to `FrozenSecurityContext`, just calling it `Frozen` feels a bit out of place.

boggle commented:
Also: Did you benchmark the changes?

Mats-SX commented:
@boggle On `Frozen`, we figured it being an inner class (`SecurityContext.Frozen`) made the name make sense. Calling it `SecurityContext.FrozenSecurityContext` is a mouthful.

If it moves outside of the interface, I agree the name needs to change.

No specific benchmark has been made for this PR. No performance changes are expected, I think.

boggle commented:
@Mats-SX ok, didn't spot `Frozen` to be inner. On benchmarking, it looked like we now always instantiate security context objects that used to be enums, did I misread that? Let's perhaps talk later.

craigtaverner commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/security/AccessModeSnapshot.java:
```diff
@@ -1,103 +0,0 @@
-/*
```
I do not find a replacement for AccessModeSnapshot and AccessModeSnapshotTest. I would think at least the test code should exist in some form?

Mats-SX replied:
`AccessModeSnapshot` is removed, because `AccessMode` is now immutable. Previously it was mutable because `AuthSubject` inherited from it. This is now handled using `SecurityContext`.

craigtaverner commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/security/AccessModeSnapshotTest.java:
```diff
@@ -1,48 +0,0 @@
-
-import static org.junit.Assert.assertEquals;
-
-public class AccessModeSnapshotTest
```
I do not find a replacement for AccessModeSnapshot and AccessModeSnapshotTest. I would think at least the test code should exist in some form?

henriknyman commented:
@Mats-SX I think some performance change is actually expected on nested tight loop procedure calls compared to the access mode snapshot code because of the additional object creation with the immutable design (also generating more garbage). This feels like a small step backwards from a performance point of view and I think we should profile it. It could easily be changed to have access mode mutable within the SecurityContext, so that all overriding is done only on access mode.

Mats-SX commented on manual/cypher/graphgist/src/main/java/org/neo4j/doc/cypherdoc/BlockType.java:
```diff
@@ -39,13 +39,15 @@
 import org.neo4j.visualization.graphviz.GraphvizWriter;
 import org.neo4j.walk.Walker;
 
+
```
This may break checkstyle.

Mats-SX committed:
Reinstate removed test

Remove empty line

fickludd committed:
Adapts TransactionData to SecurityContext

fickludd committed:
micro clean-up
